src/protocol/boolean/random_bits_generator.rs

```diff
@@ -6,28 +6,60 @@ use crate::{
     error::Error,
     ff::PrimeField,
     helpers::TotalRecords,
-    protocol::{context::Context, BasicProtocols, RecordId},
+    protocol::{context::Context, BasicProtocols, RecordId, Substep},
     secret_sharing::Linear as LinearSecretSharing,
 };
 use std::{
     marker::PhantomData,
     sync::atomic::{AtomicU32, AtomicUsize, Ordering},
 };
 
-/// A struct that pre-generates and buffers random sharings of bits from the
+/// A struct that generates random sharings of bits from the
 /// `SolvedBits` protocol. Any protocol who wish to use a random-bits can draw
 /// one by calling `generate()`.
 ///
 /// This object is safe to share with multiple threads.  It uses an atomic counter
 /// to manage concurrent accesses.
 #[derive(Debug)]
 pub struct RandomBitsGenerator<F, S, C> {
-    ctx: C,
-    record_id: AtomicU32,
+    /// Happy path touches this generator only. With the right field size, the probability of this
+    /// generator failing is very low. For [`Fp32BitPrime`] it fails with $5/2^{32}$ ~ $1^{10^-9}$
+    ///
+    /// [`Fp31`]: crate::ff::Fp32BitPrime
+    default_generator: CountingGenerator<F, S, C>,
+    /// If the default generator fails, this one is used until it successfully generates random
+    /// sharings of value `v` < [`PRIME`]
+    ///
+    /// [`PRIME`]: PrimeField::PRIME
+    fallback_generator: CountingGenerator<F, S, C>,
     abort_count: AtomicUsize,
+}
+
+/// Internal type to generate random bits using [`solved_bits`] protocol using sequential record ids.
+#[derive(Debug)]
+struct CountingGenerator<F, S, C> {
+    counter: AtomicU32,
+    ctx: C,
     _marker: PhantomData<(F, S)>,
 }
 
+enum RBGStep {
+    /// Special communication channel used when values generated using the standard communications
+    /// failed to pass the prime test. It is widely inefficient to use, so field must be large enough
```

Comment:
Wildly?

Reply:
I don't have a better word for sending 32 bytes through and waiting for response :(